### PR TITLE
Generate DNS server add/del net mgmt events

### LINKED ISSUE #16924: Add DNS server added/removed events to net_mgmt
At the moment, the net_mgmt callback system supports callbacks for when IP addresses are added or removed from interfaces but there is no way to get a notification when a DNS server (i.e. from DHCP) is added or removed which would be a useful feature so that applications know when they can safely query domain names.

This will be of most use on constrained protocols or those with high latency, such as NB-IoT, which can have a latency of 10 seconds.

The idea is to add new DNS added and DNS removed events to net_mgmt callbacks which are triggered when a server IP is added or removed.

### PR DESCRIPTION
**Pull request by jukkar**
The first L4 layer commit is also present in other PRs.
Fixes #16924

### CODE REVIEW

**Comment by pfalcon** [comment]
@lairdjm, As the original requester, can you please review this?

**Comment by pfalcon** [comment]
So, motivation of the change is not clear from the commit messages. Instead of:

> If DNS server(s) are added or removed, send proper net-mgmt
event so that we get information about the event.

I'd suggest something like:

> If DNS server(s) are added or removed (e.g., as part of DHCP processing), send newly defined net-mgmt events, so that a user application may get this information (e.g. to be aware when Internet access is "fully available").

Example and motivation is just per my understanding from cursory watching original mailing list post (maybe it needs adjustment).

**Review by nordicjm — APPROVED** [review]
Approved. I also added some tests as a pull request too which @jukkar can add to this

**Comment by jukkar** [comment]
@lairdjm I slightly modified your commit and removed the extra no-IPv6 conf file, we can have the same effect via testcase.yaml file.

**Comment by lairdjm** [comment]
@jukkar Seems I missed the else being on new line on line 149 and didn't check lengths of various lines, my bad.

**Comment by jukkar** [comment]
> Seems I missed the else being on new line on line 149 and didn't check lengths of various lines, my bad.

Yep, the problem was that if you submit to some private repo, then sanitychecker is not run. I will fix these in one go in order to proceed and we can avoid one round of changes from you.

**Comment by jukkar** [comment]
Rebased against latest master.

**Review by nordicjm — APPROVED** [review]
LGTM

**Comment by jukkar** [comment]
Rebased because L1 mgmt layer was removed in #17290 which cause merge conflict here.

File: subsys/net/lib/dns/resolve.c [hunk 1]
```diff
@@ -914,6 +927,20 @@ int dns_resolve_close(struct dns_resolve_context *ctx)
 
 	for (i = 0; i < SERVER_COUNT; i++) {
 		if (ctx->servers[i].net_ctx) {
+			struct net_if *iface;
+
+			iface = net_context_get_iface(ctx->servers[i].net_ctx);
+
+#if defined(CONFIG_NET_MGMT_EVENT_INFO)
```
Comment (tbursztyka) [inline subsys/net/lib/dns/resolve.c]: Can't you use IS_ENABLED() here? (if not, it probably needs dummies in net_mgmt.h)

Comment (jukkar) [reply subsys/net/lib/dns/resolve.c]: Good point, fixed.

**Review by tbursztyka — APPROVED** [review]

**Pull request merged**